homeassistant/components/sonos/alarms.py

```diff
@@ -0,0 +1,71 @@
+"""Class representing Sonos alarms."""
+from __future__ import annotations
+
+from collections.abc import Iterator
+import logging
+
+from pysonos import SoCo
+from pysonos.alarms import Alarm, get_alarms
+from pysonos.exceptions import SoCoException
+
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+
+from .const import DATA_SONOS, SONOS_ALARMS_UPDATED, SONOS_CREATE_ALARM
+from .household_coordinator import SonosHouseholdCoordinator
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class SonosAlarms(SonosHouseholdCoordinator):
+    """Coordinator class for Sonos alarms."""
+
+    def __init__(self, *args) -> None:
+        """Initialize the data."""
+        super().__init__(*args)
+        self._alarms: dict[str, Alarm] = {}
+
+    def __iter__(self) -> Iterator:
+        """Return an iterator for the known alarms."""
+        alarms = list(self._alarms.values())
+        return iter(alarms)
+
+    def get(self, alarm_id: str) -> Alarm | None:
+        """Get an Alarm instance."""
+        return self._alarms.get(alarm_id)
+
+    async def async_update_entities(self, soco: SoCo) -> bool:
+        """Create and update alarms entities, return success."""
+        new_alarms = await self.hass.async_add_executor_job(self.update_cache, soco)
+        if new_alarms is None:
+            return False
```

Comment:
Why do we only update existing alarms if there are new alarms found?

Reply:
Good point. Although the on/off state will work without it, we still need line 47 for attribute changes (room changes, etc).

Reply:
It will only be `None` if the API call fails. Otherwise it will be any empty `set` and pass this guard.

Reply:
It's more intuitive to catch the error here, I think. Then we can make the return value typing more strict in `update_cache`.

Reply:
Moved exception handling up a level. This also lets us discard a coordinator if it fails during `setup()` and try again with another speaker later.